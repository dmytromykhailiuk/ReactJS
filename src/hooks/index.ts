export * from "./useInputValue";
export * from "./useLocalTime";
