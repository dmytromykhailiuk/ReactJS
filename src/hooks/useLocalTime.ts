import { useState, useEffect } from "react";

const getLocalTimeString = () => new Date().toLocaleTimeString("it-IT");

export function useLocalTime(delay = 1000) {
  const [localTime, changeLocalTime] = useState(getLocalTimeString());

  useEffect(() => {
    const interval = setInterval(() => {
      changeLocalTime(getLocalTimeString());
    }, delay);

    return () => {
      clearInterval(interval);
    };
  }, []);

  return localTime;
}
