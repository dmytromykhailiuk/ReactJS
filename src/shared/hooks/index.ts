export * from "./useInputValue";
export * from "./useClickOutside";
