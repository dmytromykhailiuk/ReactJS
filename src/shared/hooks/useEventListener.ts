import { EventTypes } from "../enums/event-types";
import { useEffect } from "react";

export function useEventListener(
  event: EventTypes,
  fn: (e: any) => void,
  deps: any[],
  shouldUse: boolean = true
): void {
  useEffect(() => {
    if (shouldUse) {
      document.addEventListener(event, fn);
      return () => {
        document.removeEventListener(event, fn);
      };
    }
  }, deps);
}
