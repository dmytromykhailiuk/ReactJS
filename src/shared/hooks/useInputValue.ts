import { useState, useCallback } from "react";

export function useInputValue(message: string) {
  const [value, changeValue] = useState(message);

  const onChangeValue = useCallback((event) => {
    changeValue(event.target.value);
  }, []);

  return {
    value,
    onChangeValue,
  };
}
