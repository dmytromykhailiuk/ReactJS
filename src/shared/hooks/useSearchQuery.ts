import { useLocation } from "react-router";

export function useSearchQuery() {
  return new URLSearchParams(useLocation().search).get("SearchQuery");
}
