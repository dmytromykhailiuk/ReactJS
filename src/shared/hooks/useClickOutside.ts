import { useEffect, useMemo } from "react";
import { clickOutside } from "shared/helpers";
import { useEventListener } from "./useEventListener";
import { EventTypes } from "shared/enums";

export function useClickOutside(
  className: string,
  callback: () => void,
  shouldUse: boolean = true
) {
  const clickOutsides = useMemo(() => clickOutside(className, callback), []);

  useEventListener(EventTypes.CLICK, clickOutsides, [shouldUse], shouldUse);
}
