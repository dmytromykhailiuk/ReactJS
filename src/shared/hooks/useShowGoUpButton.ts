import { useCallback, useEffect, useState } from "react";
import { useEventListener } from "./useEventListener";
import { EventTypes } from "shared/enums";

export function useShowGoUpButton(): boolean {
  const [showGoUpButton, setShowGoUpButtonValue] = useState<boolean>(false);

  const setShowGoUpButton = useCallback(() => {
    if (!showGoUpButton && window.pageYOffset > 200) {
      setShowGoUpButtonValue(true);
    } else if (showGoUpButton && window.pageYOffset < 100) {
      setShowGoUpButtonValue(false);
    }
  }, [showGoUpButton]);

  useEventListener(EventTypes.SCROLL, setShowGoUpButton, [showGoUpButton]);

  return showGoUpButton;
}
