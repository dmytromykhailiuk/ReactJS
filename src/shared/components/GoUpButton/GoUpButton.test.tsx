import { shallow } from 'enzyme';
import React from 'react';
import toJson from 'enzyme-to-json';
import { useShowGoUpButton } from '../../hooks';
import { GoUpButton } from '.';

jest.mock('../../hooks');

describe('GoUpButton', () => {
  it('should match first snapshot', () => {
    (useShowGoUpButton as jest.Mock).mockReturnValue(true);
    const wrapper = shallow(<GoUpButton />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should contains 'scroll-btn--hide'", () => {
    (useShowGoUpButton as jest.Mock).mockReturnValue(false);
    const wrapper = shallow(<GoUpButton />);
    expect(wrapper.exists('.scroll-btn--hide')).toBeTruthy();
  });

  it('should call useShowGoUpButton function', () => {
    (useShowGoUpButton as jest.Mock).mockReturnValue(true);
    shallow(<GoUpButton />);
    expect(useShowGoUpButton).toHaveBeenCalled();
  });
});
