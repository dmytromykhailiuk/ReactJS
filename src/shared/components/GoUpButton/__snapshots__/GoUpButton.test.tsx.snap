// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoUpButton should match first snapshot 1`] = `
<div
  className="scroll-btn scroll-btn--hide"
  onClick={[Function]}
>
  <p>
    UP
  </p>
</div>
`;

exports[`GoUpButton should match second snapshot 1`] = `
<div
  className="scroll-btn"
  onClick={[Function]}
>
  <p>
    UP
  </p>
</div>
`;
